dbms/src/Storages/Page/V3/PageDirectory.cpp

```diff
@@ -1233,23 +1233,22 @@ void PageDirectory::applyRefEditRecord(
 
 void PageDirectory::apply(PageEntriesEdit && edit, const WriteLimiterPtr & write_limiter)
 {
-    // Note that we need to make sure increasing `sequence` in order, so it
-    // also needs to be protected by `write_lock` throughout the `apply`
-    // TODO: It is totally serialized, make it a pipeline
-    std::unique_lock write_lock(table_rw_mutex);
-    const UInt64 last_sequence = sequence.load();
-    // new sequence allocated in this edit
-    UInt64 new_sequence = last_sequence + 1;
+    auto edit_size = edit.getMutRecords().size();
+    auto my_sequence = sequence.fetch_add(edit_size);
```

Comment:
Adding `PageDirectory.sequence` out of order will bring bugs

Assume that
- Original status: PageDirectory with page_id = 100, sequence=222
- write thread A with delete 100
- read thread B that accessing to page_id=100

Thread A enter `PageDirectory::apply` and run into line 1237, the sequence is increased to 223
Thread B create a snapshot with `sequence=223`. But it currently can read page_id=100.
Thread A resume and continue `apply`, delete page 100
Thread B find it can not read page 100 with same page_id=100